### PR TITLE
Use venv API to get venv paths

### LINKED ISSUE #701: pipx enters in infinite loop when trying to install a package on python distributed with MSYS2 on Windows
**Describe the bug**
When using MSYS2, package mingw-w64-x86_64-python (https://packages.msys2.org/package/mingw-w64-x86_64-python), pipx enters in infinite loop when installing a package.

```
pipx install pycowsay --verbose
pipx >(setup:717): pipx version is 0.16.3
pipx >(setup:718): Default python interpreter is 'c:/users/user1/software/tools/msys64/mingw64/bin/python3.exe'
pipx >(package_name_from_spec:323): Determined package name: pycowsay
pipx >(package_name_from_spec:324): Package name determined in 0.0s
creating virtual environment...
pipx >(run_subprocess:135): running c:/users/user1/software/tools/msys64/mingw64/bin/python3.exe -m venv --without-pip C:/Users/user1/.local/pipx/venvs/pycowsay
creating shared libraries...
pipx >(run_subprocess:135): running c:/users/user1/software/tools/msys64/mingw64/bin/python3.exe -m venv --clear C:/Users/user1/.local/pipx/shared
creating shared libraries...
pipx >(run_subprocess:135): running c:/users/user1/software/tools/msys64/mingw64/bin/python3.exe -m venv --clear C:/Users/user1/.local/pipx/shared
creating shared libraries...
pipx >(run_subprocess:135): running c:/users/user1/software/tools/msys64/mingw64/bin/python3.exe -m venv --clear C:/Users/user1/.local/pipx/shared
creating shared libraries...
pipx >(run_subprocess:135): running c:/users/user1/software/tools/msys64/mingw64/bin/python3.exe -m venv --clear C:/Users/user1/.local/pipx/shared
creating shared libraries...
pipx >(run_subprocess:135): running c:/users/user1/software/tools/msys64/mingw64/bin/python3.exe -m venv --clear C:/Users/user1/.local/pipx/shared
creating shared libraries...
pipx >(run_subprocess:135): running c:/users/user1/software/tools/msys64/mingw64/bin/python3.exe -m venv --clear C:/Users/user1/.local/pipx/shared
(...)
```

if I execute manually the command `python3.exe -m venv --clear C:/Users/user1/.local/pipx/shared` it is executed successfully

```
c:/users/user1/software/tools/msys64/mingw64/bin/python3.exe -m venv --clear C:/Users/user1/.local/pipx/shared
Looking in links: c:\Users\user1\AppData\Local\Temp\tmphnr44wk9
Processing c:\users\user1\appdata\local\temp\tmphnr44wk9\setuptools-49.2.1-py3-none-any.whl
Processing c:\users\user1\appdata\local\temp\tmphnr44wk9\pip-20.2.3-py2.py3-none-any.whl
Installing collected packages: setuptools, pip
Successfully installed pip-20.2.3 setuptools-49.2.1
```

Contents of pipx logs:

```
   381.0ms (setup:715): 2021-06-02 20:15:15
   381.0ms (setup:716): C:\Users\user1\.local\bin\pipx install pycowsay
   381.0ms (setup:717): pipx version is 0.16.3
   381.0ms (setup:718): Default python interpreter is 'c:/users/user1/software/tools/msys64/mingw64/bin/python3.exe'
   383.0ms (package_name_from_spec:323): Determined package name: pycowsay
   383.0ms (package_name_from_spec:324): Package name determined in 0.0s
   385.0ms (run_subprocess:135): running c:/users/user1/software/tools/msys64/mingw64/bin/python3.exe -m venv --without-pip C:/Users/user1/.local/pipx/venvs/pycowsay
   843.1ms (run_subprocess:148): stdout:
   844.1ms (run_subprocess:150): stderr:
   844.1ms (run_subprocess:151): returncode: 0
   846.1ms (run_subprocess:135): running c:/users/user1/software/tools/msys64/mingw64/bin/python3.exe -m venv --clear C:/Users/user1/.local/pipx/shared
 16771.8ms (run_subprocess:148): stdout: Looking in links: c:\Users\user1\AppData\Local\Temp\tmpxbeu6b31
Processing c:\users\user1\appdata\local\temp\tmpxbeu6b31\setuptools-49.2.1-py3-none-any.whl
Processing c:\users\user1\appdata\local\temp\tmpxbeu6b31\pip-20.2.3-py2.py3-none-any.whl
Installing collected packages: setuptools, pip
Successfully installed pip-20.2.3 setuptools-49.2.1
 16771.8ms (run_subprocess:150): stderr:
 16772.8ms (run_subprocess:151): returncode: 0
 16774.8ms (run_subprocess:135): running c:/users/user1/software/tools/msys64/mingw64/bin/python3.exe -m venv --clear C:/Users/user1/.local/pipx/shared
 34595.2ms (run_subprocess:148): stdout: Looking in links: c:\Users\user1\AppData\Local\Temp\tmpreoye3oe
Processing c:\users\user1\appdata\local\temp\tmpreoye3oe\setuptools-49.2.1-py3-none-any.whl
Processing c:\users\user1\appdata\local\temp\tmpreoye3oe\pip-20.2.3-py2.py3-none-any.whl
Installing collected packages: setuptools, pip
Successfully installed pip-20.2.3 setuptools-49.2.1
 34595.2ms (run_subprocess:150): stderr:
 34595.2ms (run_subprocess:151): returncode: 0
 34597.2ms (run_subprocess:135): running c:/users/user1/software/tools/msys64/mingw64/bin/python3.exe -m venv --clear C:/Users/user1/.local/pipx/shared
 51837.0ms (run_subprocess:148): stdout: Looking in links: c:\Users\user1\AppData\Local\Temp\tmp_75yare0
Processing c:\users\user1\appdata\local\temp\tmp_75yare0\setuptools-49.2.1-py3-none-any.whl
Processing c:\users\user1\appdata\local\temp\tmp_75yare0\pip-20.2.3-py2.py3-none-any.whl
Installing collected packages: setuptools, pip
Successfully installed pip-20.2.3 setuptools-49.2.1
 51838.0ms (run_subprocess:150): stderr:
 51838.0ms (run_subprocess:151): returncode: 0
 (
 (...)
```

I have seen a similar issue happen with this particular python distribution (MSYS2) and other tools. For example  https://github.com/python-poetry/poetry/issues/2867

This distribution has one particularity. Although it is built for Windows, venvs created with it have POSIX-style:

```
user1@HOST ~/A/L/p/C/v/somevenv> ls -al
total 5
drwxr-xr-x 1 user1 user1   0 Aug 29 23:05 ./
drwxr-xr-x 1 user1 user1   0 Aug 29 23:05 ../
drwxr-xr-x 1 user1 user1   0 Aug 29 23:06 bin/
drwxr-xr-x 1 user1 user1   0 Aug 29 23:05 include/
drwxr-xr-x 1 user1 user1   0 Aug 29 23:05 lib/
-rw-r--r-- 1 user1 user1 106 Aug 29 23:05 pyvenv.cfg
````

The venvs created with have `lib` and `bin` directories instead of `Scripts` directories.  I suspect that pipx may be expecting a Windows-style venv because the underlying OS is Windows. See https://github.com/pypa/pipx/blob/5e26b3b857a568be7fde40a9ef2caa8f7c0c040d/src/pipx/util.py#L79

it is possible to detect this python distro with sysconfig.get_platform() . It will return mingw . See https://stackoverflow.com/questions/35956180/how-to-determine-the-python-is-mingw-or-official-build

**How to reproduce**

1 - Install MSYS2 on Windows 10
2 - From MSYS shell, install python mingw64 : `pacman -S mingw-w64-x86_64-python` and `pacman -S mingw-w64-x86_64-python-pip`
3 - Install pipx using PIP
4 - `pipx install pycowsay`
5 - The process will never complete and venv will be cleaned and rebuilt in an infinite loop

**Expected behavior**

Install should complete successfully

### PR DESCRIPTION
**Pull request by meowmeowmeowcat**
* [ ] I have added an entry to `docs/changelog.md`

## Summary of changes

Use venv API to get venv paths.

Closes #701 

## Test plan

Tested by running
```
pipx install black --verbose
```

### CODE REVIEW

**Comment by meowmeowmeowcat** [comment]
However, some users might not have venv installed (see #853). Maybe we should also support virtualenv?

File: src/pipx/util.py [hunk 1]
```diff
@@ -112,19 +112,26 @@ def run_pypackage_bin(bin_path: Path, args: List[str]) -> NoReturn:
...
-
-    def get_venv_paths(root: Path) -> Tuple[Path, Path]:
-        bin_path = root / "bin"
-        python_path = bin_path / "python"
-        return bin_path, python_path
+def get_venv_paths(python: str, root: Path) -> Tuple[Path, Path]:
+    command_str = textwrap.dedent(
+        f"""
+        import venv
+        root_paths = venv.EnvBuilder().ensure_directories(r"{str(root)}")
```
Comment (uranusjr) [inline src/pipx/util.py:119]:
```suggestion
        root_paths = venv.EnvBuilder().ensure_directories({root})
```

**Review by uranusjr — COMMENTED** [review]
I have a gut feeling this may be problematic if the environment path contains non-ASCII characters, since not all terminals can output those characters (say if the encoding is not configured to UTF-8).

It’s probably more reobust to write the path to a temp file instead.

**Comment by uranusjr** [comment]
> Maybe we should also support virtualenv?

Does virtualenv provide an interface to introspect an environment _outside of it_? Because an environment created by virtualenv does not have access to the virtualenv module.

File: src/pipx/util.py (hunk 1 above)
Comment (meowmeowmeowcat) [reply src/pipx/util.py:119]: Probably raw string should be used? (For handling paths on Windows)

**Comment by meowmeowmeowcat** [comment]
> I have a gut feeling this may be problematic if the environment path contains non-ASCII characters, since not all terminals > can output those characters (say if the encoding is not configured to UTF-8).
>
> It’s probably more reliable to write the path to a temp file instead.

https://github.com/pypa/pipx/blob/4c0f2892c8f932f345a379b84bffe816c7c7206c/src/pipx/util.py#L130-L137
The output is just printed to the terminal in this function and it seems fine?

**Comment by meowmeowmeowcat** [comment]
> Does virtualenv provide an interface to introspect an environment outside of it?

I think no. Or probably we can keep this for virtualenv users?
```py
if WINDOWS:

    def get_venv_paths(root: Path) -> Tuple[Path, Path]:
        bin_path = root / "Scripts"
        python_path = bin_path / "python.exe"
        return bin_path, python_path

else:

    def get_venv_paths(root: Path) -> Tuple[Path, Path]:
        bin_path = root / "bin"
        python_path = bin_path / "python"
        return bin_path, python_path
```

**Comment by uranusjr** [comment]
If we need to keep the `bin`/`Script` guess logic I’d rather keep it and improve the detection logic to account for MSYS. I think there’s established logic out there somewhere that can reliably detect this.

**Comment by meowmeowmeowcat** [comment]
> I think there’s established logic out there somewhere that can reliably detect this.

Yes, for example https://github.com/python-poetry/poetry/pull/4467 and https://github.com/python-poetry/poetry/pull/3713.

**Comment by meowmeowmeowcat** [comment]
> If we need to keep the bin/Script guess logic I’d rather keep it and improve the detection logic to account for MSYS.

I will open another PR for this.

**Pull request closed without merging**